File: 2020/day_18/day_18.py
with open('in') as f:
    data = list(map(lambda l: list(l.strip().replace(' ', '')), f.readlines()))


def evaluate(e):
    value = None
    operation = None

    i = 0
    while i < len(e):
        c = e[i]

        if c in '+*':
            operation = c
        elif c in '0123456789':
            c = int(c)
            if operation is None:
                value = c
            else:
                if operation == '+':
                    value += c
                else:
                    value *= c
        elif c == '(':
            tmp_e = []
            i += 1

            level = 1
            while True:
                if e[i] == '(':
                    level += 1
                elif e[i] == ')':
                    level -= 1

                if level == 0:
                    break
                else:
                    tmp_e.append(e[i])

                i += 1

            b = evaluate(tmp_e)
            if operation is None:
                value = b
            else:
                if operation == '+':
                    value += b
                else:
                    value *= b
        else:
            assert False

        i += 1

    return value


def evaluate_2(e):
    i = 0
    while i < len(e):
        if e[i] == '+':
            level = 0
            index = -1
            for j in range(i - 1, -1, -1):
                if e[j] == ')':
                    level += 1
                elif e[j] == '(':
                    level -= 1

                if (level == 0 and e[j] in '+*') or (level < 0 and e[j] in '()'):
                    index = j
                    break
            e.insert(index + 1, '(')
            i += 1

            level = 0
            index = len(e)
            for j in range(i + 1, len(e)):
                if e[j] == '(':
                    level += 1
                elif e[j] == ')':
                    level -= 1

                if (level == 0 and e[j] in '+*') or (level < 0 and e[j] in '()'):
                    index = j
                    break
            e.insert(index, ')')

        i += 1

    return evaluate(e)


def part_one():
    return sum(map(evaluate, data))


def part_two():
    return sum(map(evaluate_2, data))


if __name__ == '__main__':
    print(f'Part One: {part_one()}')
    print(f'Part Two: {part_two()}')

File: 2020/day_18/day_18.rs
use std::fs;

#[derive(Debug)]
enum Expression {
    Constant(u64),
    Plus(Box<Expression>, Box<Expression>),
    Times(Box<Expression>, Box<Expression>),
}

impl Expression {
    fn next_op(e: &str, start: usize) -> Option<usize> {
        let mut level = 0;
        for (i, c) in e.chars().enumerate() {
            if c == '(' {
                level += 1;
            } else if c == ')' {
                level -= 1;
            }

            if level == 0 && (c == '+' || c == '*') {
                return Some(i + start);
            }
        }

        None
    }

    fn operator_precedance(op: &str, plus_precedance: i32, times_precedance: i32) -> i32 {
        match op {
            "+" => plus_precedance,
            "*" => times_precedance,
            _ => panic!("Invalid operator"),
        }
    }

    fn from_str(mut e: &str, plus_precedance: i32, times_precedance: i32) -> Expression {
        // Check if constant.
        if let Ok(v) = e.parse() {
            return Expression::Constant(v);
        }

        // Remove unneeded parenthesis.
        let mut clean = false;
        while !clean {
            let mut level = 0;
            for (i, c) in e.chars().enumerate() {
                if c == '(' {
                    level += 1;
                } else if c == ')' {
                    level -= 1;
                }

                if level == 0 && i != (e.len() - 1) {
                    clean = true;
                    break;
                }
            }

            if !clean {
                e = &e[1..(e.len() - 1)];
            }
        }

        // Find operator.
        let mut operator = Expression::next_op(e, 0).unwrap();
        while let Some(next_operator) = Expression::next_op(&e[(operator + 1)..], operator + 1) {
            let current_precedance = Expression::operator_precedance(
                &e[operator..(operator + 1)],
                plus_precedance,
                times_precedance,
            );
            let next_precedance = Expression::operator_precedance(
                &e[next_operator..(next_operator + 1)],
                plus_precedance,
                times_precedance,
            );

            if next_precedance <= current_precedance {
                operator = next_operator;
            } else {
                break;
            }
        }

        let op = &e[operator..(operator + 1)];
        let lhs = Expression::from_str(&e[..operator], plus_precedance, times_precedance);
        let rhs = Expression::from_str(&e[(operator + 1)..], plus_precedance, times_precedance);
        match op {
            "*" => Expression::Times(Box::new(lhs), Box::new(rhs)),
            "+" => Expression::Plus(Box::new(lhs), Box::new(rhs)),
            _ => panic!("Invalid operator"),
        }
    }

    fn eval(&self) -> u64 {
        match self {
            Expression::Constant(x) => *x,
            Expression::Times(lhs, rhs) => lhs.eval() * rhs.eval(),
            Expression::Plus(lhs, rhs) => lhs.eval() + rhs.eval(),
        }
    }
}

fn main() {
    let data: Vec<_> = fs::read_to_string("in")
        .unwrap()
        .lines()
        .map(|l| l.trim().replace(" ", ""))
        .collect();

    println!("Part One: {}", part_one(&data));
    println!("Part Two: {}", part_two(&data));
}

fn part_one(data: &Vec<String>) -> u64 {
    data.iter()
        .map(|line| Expression::from_str(line, 1, 1).eval())
        .sum()
}

fn part_two(data: &Vec<String>) -> u64 {
    data.iter()
        .map(|line| Expression::from_str(line, 2, 1).eval())
        .sum()
}
